soc: rp2350: Add basic Asymmetric Multiprocessing (AMP) support

dancollins opened the pull request:
I have a project using the RP2350, and figured it would be cool to run Zephyr on both cores. I have got a basic sample working that has both cores running Zephyr and printing a message on their respective UARTs.

I have noticed the reset controller, as an example, makes use of read-modify-writes that will not be compatible with my AMP setup. This, along with other peripherals, will need to either have hardware locking (RP2350 includes a hardware spinlock) or atomic access (RP2350 has memory addresses for atomic set/clear).

No doubt issues will crop up as I use this implementation - so I will keep it in draft status until I stop needing to make any changes to it. I intend on being able to demonstrate an OpenAMP sample running, as this should provide all the basic functionality needed for AMP.

josuah commented:
The CI [error message](https://github.com/zephyrproject-rtos/zephyr/actions/runs/15379004468/job/43267455320?pr=90922#step:12:783):

```
/__w/zephyr/modules/hal/rpi_pico/src/rp2_common/hardware_base/include/hardware/address_mapped.h:124:27: error: implicit declaration of function 'typeof' [-Werror=implicit-function-declaration]
  124 | #define hw_xor_alias(p) ((typeof(p))hw_xor_alias_untyped(p))
      |                           ^~~~~~
/__w/zephyr/modules/hal/rpi_pico/src/rp2_common/hardware_flash/flash.c:302:6: note: in expansion of macro 'hw_xor_alias'
  302 |     *hw_xor_alias(devinfo) = (*devinfo ^ wdata) & mask;
      |      ^~~~~~~~~~~~
/__w/zephyr/modules/hal/rpi_pico/src/rp2_common/hardware_base/include/hardware/address_mapped.h:124:27: error: called object is not a function or function pointer
  124 | #define hw_xor_alias(p) ((typeof(p))hw_xor_alias_untyped(p))
      |                          ~^~~~~~~~~~
/__w/zephyr/modules/hal/rpi_pico/src/rp2_common/hardware_flash/flash.c:302:6: note: in expansion of macro 'hw_xor_alias'
  302 |     *hw_xor_alias(devinfo) = (*devinfo ^ wdata) & mask;
      |      ^~~~~~~~~~~~
cc1: all warnings being treated as errors
ninja: build stopped: subcommand failed.
```

It looks like C language features used in RPi SDK are getting in the way.

dsseng reviewed (commented):
just some thoughts

dancollins committed:
drivers: pi_pico: add ipm drivers

This commit introduces an interprocess mailbox driver
for the Raspberry Pi Pico.

Signed-off-by: Dan Collins <dan@collinsnz.com>

dancollins committed:
soc: rp2350: integrate support for second CPU core

This commit introduces a means to build Zephyr for both
Cortex M33 cores of the RP2350, and for CPU0 to boot CPU1.
It includes a partition table for the rpi_pico2 board that
demonstrates how the two cores can operate independently.

Signed-off-by: Dan Collins <dan@collinsnz.com>

dancollins committed:
soc: rp2350: skip clock initialisation for cpu1

We don't want to reinitialise the clock control when
we boot the second core - so this commit introduces a
means to skip initialisation (i.e. assume it is already
initialised).

Signed-off-by: Dan Collins <dan@collinsnz.com>

dancollins commented:
I have split this out into three commits for now - this makes it easier to track what's going on. I've also removed the changes to the various samples/tests as I really like @dsseng's idea to avoid forcing the `_cpu0` label onto all the non-AMP builds.

I also like the idea of using the device tree for disabling the clock init, rather than the Kconfig symbol I have. I'm not sure how to do that exactly - whether it's a special compat for the second core, or just a flag.

dsseng commented:
As we discussed with @dancollins on Discord, I'll do some work, including implementing my proposed changes and perhaps IPM on my branch and notify here when some major progress is there

dsseng commented:
Here you can find some work that have been done by me:

- https://github.com/dsseng/zephyr/commits/for-dan/amp-rp2350/
- https://github.com/dsseng/zephyr/compare/9a579826d78ba02826a04a1b6a56beaac061b59d..b8bfa28b2127bb93676d25f1080eed9d138f6cf3 (only previously unsubmitted parts)

This branch includes the following PRs:
- https://github.com/zephyrproject-rtos/zephyr/pull/91200 - necessary
- https://github.com/zephyrproject-rtos/zephyr/pull/92321 - might be not necessary, but fixes stability issues for some devices
- https://github.com/zephyrproject-rtos/zephyr/pull/92654 - needed to make use of `ipc` samples

TODO:
- [a whitespace-only change fell in accidentally](https://github.com/dsseng/zephyr/compare/9a579826d78ba02826a04a1b6a56beaac061b59d..b8bfa28b2127bb93676d25f1080eed9d138f6cf3#diff-1f974f04f0e9463cdb334bf183d3d86d94160a022f854085f57ffc878fb7de06)
- Search for a better option for skipping fundamental clocks on CPU1
- (might leave this for later) Make some hardware like clock, reset, maybe GPIO more shareable by making sure they use atomic accesses. When confirmed, add such peripherals to both dts.

Samples to try:
- No IPC, just separate apps with assigned peripherals (needs https://github.com/zephyrproject-rtos/zephyr/pull/91905)
  - `west build -d build/0 -b rpi_pico2/rp2350a/m33_cpu0 samples/basic/blinky --pristine=auto -- -DCONFIG_XIP=n && west build -d build/1 -b rpi_pico2/rp2350a/m33_cpu1 -S video-sw-generator samples/subsys/usb/uvc --pristine=auto && picotool load -nuv -t elf ./build/1/zephyr/zephyr.elf && picotool load -nuvx -t elf ./build/0/zephyr/zephyr.elf`
- `west build --sysbuild -b rpi_pico2/rp2350a/m33_cpu0 samples/subsys/ipc/ipc_service/icmsg --pristine=auto -- -DCONFIG_XIP=n -DDTC_OVERLAY_FILE=boards/rpi_pico2_rp2350a_m33_cpu0_icbmsg.overlay -Dremote_DTC_OVERLAY_FILE=boards/rpi_pico2_rp2350a_m33_cpu1_icbmsg.overlay && picotool load -nuv -t elf ./build/remote/zephyr/zephyr.elf && picotool load -nuvx -t elf ./build/icmsg/zephyr/zephyr.elf`
- `west build --sysbuild -b rpi_pico2/rp2350a/m33_cpu0 samples/subsys/ipc/ipc_service/icmsg --pristine=auto -- -DCONFIG_XIP=n && picotool load -nuv -t elf ./build/remote/zephyr/zephyr.elf && picotool load -nuvx -t elf ./build/icmsg/zephyr/zephyr.elf`
- `west build --sysbuild -b rpi_pico2/rp2350a/m33_cpu0 samples/subsys/ipc/ipc_service/multi_endpoint --pristine=auto -- -DCONFIG_XIP=n -DDTC_OVERLAY_FILE=boards/rpi_pico2_rp2350a_m33_cpu0_icbmsg.overlay -Dremote_DTC_OVERLAY_FILE=boards/rpi_pico2_rp2350a_m33_cpu1_icbmsg.overlay && picotool load -nuv -t elf ./build/remote/zephyr/zephyr.elf && picotool load -nuvx -t elf ./build/multi_endpoint/zephyr/zephyr.elf`
- `west build --sysbuild -b rpi_pico2/rp2350a/m33_cpu0 samples/subsys/ipc/ipc_service/multi_endpoint --pristine=auto -- -DCONFIG_XIP=n && picotool load -nuv -t elf ./build/remote/zephyr/zephyr.elf && picotool load -nuvx -t elf ./build/multi_endpoint/zephyr/zephyr.elf`
